src/batHeap.ml

```diff
@@ -107,35 +107,41 @@ let find_min bh = match bh.mind with
 *)
 
 
-let rec find_min_tree ts k = match ts with
-  | [] -> failwith "find_min_tree"
-  | [t] -> k t
-  | t :: ts ->
-    find_min_tree ts begin
-      fun u ->
-        if Pervasives.compare t.root u.root <= 0
-        then k t else k u
-    end
-
-let rec del_min_tree bts k = match bts with
-  | [] -> invalid_arg "del_min"
-  | [t] -> k t []
-  | t :: ts ->
-    del_min_tree ts begin
-      fun u uts ->
-        if Pervasives.compare t.root u.root <= 0
-        then k t ts
-        else k u (t :: uts)
-    end
-
-let del_min bh =
-  del_min_tree bh.data begin
-    fun bt data ->
+let del_min =
+  let rec del_min_tree bts k =
+    match bts with
+      | [] ->
+          invalid_arg "del_min"
+      | [t] ->
+          k t []
+      | t :: ts ->
+          del_min_tree ts (fun u uts ->
+            if Pervasives.compare t.root u.root <= 0 then
+              k t ts
+            else
+              k u (t :: uts))
+  in
+  let rec find_min_tree ts k =
+    match ts with
+      | [] ->
+          invalid_arg "del_min"
+      | [t] ->
+          k t
+      | t :: ts ->
+          find_min_tree ts (fun u ->
+            if Pervasives.compare t.root u.root <= 0 then k t else k u)
```

Comment:
I'm not convinced by this change. I would rather have `find_mind_tree` remain a generic operation (it could be used by other functions later, or even expose to the user without the continuation-taking interface), but take two continuations, `k : 'a t -> 'b` and `kfail : unit -> 'b`, instead of one -- this is standard in functional encoding of search procedures. Then `del_min_tree` could use `(fun () -> raise "del_min")` as the failure continuation.